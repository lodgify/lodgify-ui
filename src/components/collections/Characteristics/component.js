import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'semantic-ui-react';

import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { getFirstFourItems } from 'utils/get-first-four-items';
import { Grid } from 'layout/Grid';
import { GridColumn } from 'layout/GridColumn';
import { Icon } from 'elements/Icon';
import { ShowOn } from 'layout/ShowOn';

/**
 * The standard widget for displaying a collection of characteristics for a property or a room.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ spaceCharacteristics }) => (
  <Grid columns={1}>
    <GridColumn width={12}>
      <ShowOn
        computer
        parent={List}
        parentProps={{ horizontal: true }}
        tablet
        widescreen
      >
        {getFirstFourItems(spaceCharacteristics).map(
          ({ iconName, text }, index) => (
            <List.Item key={buildKeyFromStrings(text, index)}>
              <Icon labelText={text} name={iconName} />
            </List.Item>
          )
        )}
      </ShowOn>
      <ShowOn mobile parent={Grid} parentProps={{ columns: 1 }}>
        {getFirstFourItems(spaceCharacteristics).map(
          ({ iconName, text }, index) => (
            <GridColumn
              computer={3}
              key={buildKeyFromStrings(text, index)}
              mobile={6}
            >
              <Icon labelText={text} name={iconName} />
            </GridColumn>
          )
        )}
      </ShowOn>
    </GridColumn>
  </Grid>
);

Component.displayName = 'Characteristics';

Component.defaultProps = {
  spaceCharacteristics: [],
};

Component.propTypes = {
  /** The main characteristics to display. Maximum four. */
  spaceCharacteristics: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The name of the icon to display.
       * [See here for the full list of valid icon names](https://github.com/lodgify/lodgify-ui/blob/master/src/components/elements/Icon/constants.js)
       */
      iconName: PropTypes.string.isRequired,
      /** A visible label to display for the key fact. */
      text: PropTypes.string.isRequired,
    })
  ),
};
