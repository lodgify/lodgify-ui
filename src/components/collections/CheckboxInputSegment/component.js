import React from 'react';
import PropTypes from 'prop-types';
import { Segment } from 'semantic-ui-react';

import { Divider } from 'elements/Divider';
import { FlexContainer } from 'layout/FlexContainer';
import { Grid } from 'layout/Grid';
import { Heading } from 'typography/Heading';
import { Checkbox } from 'inputs/Checkbox';
import { GridColumn } from 'layout/GridColumn';
import { withResponsive } from 'utils/with-responsive';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';

import { getGridColumns } from './utils/getGridColumns';

/**
 * A collection of information alongside checkboxes.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
const Component = ({ heading, checkboxes, isUserOnMobile }) => (
  <Segment className="is-input-segment is-checkbox-input-segment" vertical>
    <FlexContainer
      alignItems="baseline"
      flexDirection="column"
      justifyContent="space-between"
    >
      <Heading>{heading}</Heading>
      <Divider size="small" />
      <Grid columns={getGridColumns(isUserOnMobile)}>
        {checkboxes.map(({ name, label, ...checkboxProps }) => (
          <GridColumn key={buildKeyFromStrings(label, name)}>
            <Checkbox
              label={label}
              name={name}
              {...checkboxProps}
              isFluid={isUserOnMobile}
              isLabelLeft={isUserOnMobile}
            />
          </GridColumn>
        ))}
      </Grid>
    </FlexContainer>
  </Segment>
);

Component.displayName = 'CheckboxInputSegment';

Component.propTypes = {
  /** The checkboxes to display. See [the `Checkbox` component for valid props](#/Inputs/Checkbox).  */
  checkboxes: PropTypes.arrayOf(PropTypes.object).isRequired,
  /** The text for the input segment's heading. */
  heading: PropTypes.string.isRequired,
  /**
   * Is the user on a mobile device.
   * Provided by `withResponsive` so ignored in the styleguide.
   * @ignore
   */
  isUserOnMobile: PropTypes.bool.isRequired,
};

export const ComponentWithResponsive = withResponsive(Component);
