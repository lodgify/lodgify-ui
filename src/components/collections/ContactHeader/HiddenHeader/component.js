import React, { useRef, useState, useEffect } from 'react';
import {
  string,
  arrayOf,
  shape,
  func,
  bool,
  oneOfType,
  number,
} from 'prop-types';
import { Menu, Accordion } from 'semantic-ui-react';
import { size } from 'lodash';

import { Heading } from 'typography/Heading';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { Modal } from 'elements/Modal';
import { Submenu } from 'elements/Submenu';
import { testidFactory } from 'utils/testid';

import { PHONE_PREFIX } from '../constants';

const TEST_ID_PREFIX = 'hiddenHeader';

const testid = testidFactory(TEST_ID_PREFIX);

export const Component = ({
  headingHref,
  headingText,
  navigationItems,
  languageOptions,
  onChangeLanguage,
  languageValue,
  currencyOptions,
  currencyNoResultsText,
  onChangeCurrency,
  currencyValue,
  phoneNumber,
}) => {
  const [activeItem, setActiveItem] = useState('');

  useEffect(() => {
    /* istanbul ignore next */
    if (navigationItems.length > 0) {
      setActiveItem(navigationItems[0].text);
    }
  }, [navigationItems.length]);

  const menu = useRef();

  const onMenuLinkClick = text => {
    setActiveItem(text);
  };

  return (
    <Modal
      className="hidden-contact-header-modal"
      closeIcon={<Icon labelText="Close" name="close" />}
      data-testid={testid()}
      header={
        <Menu.Item href={headingHref} link>
          <Heading className="header-title" size="small">
            {headingText}
          </Heading>
        </Menu.Item>
      }
      isFullscreen
      trigger={<Icon name={ICON_NAMES.BARS} />}
    >
      <div ref={menu}>
        <Menu text vertical>
          {navigationItems.map(({ subItems, text, link }, index) =>
            size(subItems) ? (
              <Accordion
                as={Menu.Item}
                key={buildKeyFromStrings(text, index)}
                panels={[
                  {
                    title: {
                      content: text,
                      key: buildKeyFromStrings(text, index),
                    },
                    content: {
                      content: subItems.map(({ text, href }, index) => (
                        <Menu.Item
                          active={text === activeItem}
                          data-testid={testid(`accordionItem_${index}`)}
                          href={href}
                          key={buildKeyFromStrings(text, index)}
                          link
                          onClick={() => onMenuLinkClick(text)}
                        >
                          {text}
                        </Menu.Item>
                      )),
                      key: buildKeyFromStrings(index, text),
                    },
                  },
                ]}
              />
            ) : (
              <Menu.Item
                active={text === activeItem}
                data-testid={testid(`hiddenHeaderItem_${index}`)}
                href={link}
                key={buildKeyFromStrings(text, index)}
                link
                onClick={() => onMenuLinkClick(text)}
              >
                {text}
              </Menu.Item>
            )
          )}
        </Menu>
        <Menu vertical>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
      </div>
    </Modal>
  );
};

Component.defaultProps = {
  currencyNoResultsText: '',
  currencyOptions: [],
  currencyValue: null,
  headingHref: '',
  headingText: '',
  languageOptions: [],
  languageValue: null,
  phoneNumber: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The link for the heading. */
  headingHref: string,
  /** The text for the heading. */
  headingText: string,
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The items for a user to navigate the site. */
  navigationItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
};
