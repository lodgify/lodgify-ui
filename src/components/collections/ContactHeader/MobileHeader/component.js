import React, { useState, useEffect } from 'react';
import {
  string,
  arrayOf,
  shape,
  func,
  bool,
  oneOfType,
  number,
  oneOf,
} from 'prop-types';
import { Menu, Accordion } from 'semantic-ui-react';
import { size } from 'lodash';

import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { Modal } from 'elements/Modal';
import { Submenu } from 'elements/Submenu';
import { testidFactory } from 'utils/testid';
import { Logo } from 'elements/Logo';

import { PHONE_PREFIX } from '../constants';

const TEST_ID_PREFIX = 'MobileHeader';

const testid = testidFactory(TEST_ID_PREFIX);

export const Component = ({
  currencyNoResultsText,
  currencyOptions,
  currencyValue,
  languageOptions,
  languageValue,
  logoHref,
  logoSize,
  logoSizes,
  logoSrc,
  logoSrcSet,
  logoSubText,
  logoText,
  navigationItems,
  onChangeCurrency,
  onChangeLanguage,
  phoneNumber,
}) => {
  const [activeItem, setActiveItem] = useState('');

  useEffect(() => {
    /* istanbul ignore next */
    if (navigationItems.length > 0) {
      setActiveItem(navigationItems[0].text);
    }
  }, [navigationItems.length]);

  const onMenuLinkClick = text => {
    setActiveItem(text);
  };

  return (
    <Modal
      className="hidden-contact-header-modal"
      closeIcon={<Icon labelText="Close" name="close" />}
      data-testid={testid()}
      header={
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
        />
      }
      isFullscreen
      trigger={<Icon name={ICON_NAMES.BARS} />}
    >
      <div>
        <Menu text vertical>
          {navigationItems.map(({ subItems, text, href }, index) =>
            size(subItems) ? (
              <Accordion
                as={Menu.Item}
                key={buildKeyFromStrings(text, index)}
                panels={[
                  {
                    title: {
                      content: text,
                      key: buildKeyFromStrings(text, index),
                    },
                    content: {
                      content: subItems.map(({ text, href }, index) => (
                        <Menu.Item
                          active={text === activeItem}
                          data-testid={testid(`accordionItem_${index}`)}
                          href={href}
                          key={buildKeyFromStrings(text, index)}
                          link
                          onClick={() => onMenuLinkClick(text)}
                        >
                          {text}
                        </Menu.Item>
                      )),
                      key: buildKeyFromStrings(index, text),
                    },
                  },
                ]}
              />
            ) : (
              <Menu.Item
                active={text === activeItem}
                data-testid={testid(`MobileHeaderItem_${index}`)}
                href={href}
                key={buildKeyFromStrings(text, index)}
                link
                onClick={() => onMenuLinkClick(text)}
              >
                {text}
              </Menu.Item>
            )
          )}
        </Menu>
        <Menu vertical>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
      </div>
    </Modal>
  );
};

Component.defaultProps = {
  currencyNoResultsText: '',
  currencyOptions: [],
  currencyValue: null,
  languageOptions: [],
  languageValue: null,
  logoHref: '',
  logoSize: 'medium',
  logoSizes: undefined,
  logoSrc: undefined,
  logoSrcSet: undefined,
  logoSubText: '',
  logoText: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
  phoneNumber: '',
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The href for the logo link. */
  logoHref: string,
  /** The maximum size of the headers logo. */
  logoSize: oneOf(['medium', 'large', 'huge']),
  /** A list of one or more strings separated by commas indicating a set of source sizes for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSizes: string,
  /** The src url for the logo. */
  logoSrc: string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSrcSet: string,
  /** The text that appears under the logo or logo text. */
  logoSubText: string,
  /** The text for the logo. */
  logoText: string,
  /** The items for a user to navigate the site. */
  /** The navigation items displayed in the header. */
  navigationItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
};
