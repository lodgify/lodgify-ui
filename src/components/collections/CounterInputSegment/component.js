import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Segment } from 'semantic-ui-react';

import { Counter } from 'inputs/Counter';
import { Divider } from 'elements/Divider';
import { FlexContainer } from 'layout/FlexContainer';
import { Heading } from 'typography/Heading';
import { Paragraph } from 'typography/Paragraph';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';

/**
 * A collection of information alongside a counter inputs.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ heading, counters }) => (
  <Segment className="is-input-segment is-counter-input-segment" vertical>
    <FlexContainer
      alignItems="baseline"
      flexDirection="column"
      justifyContent="space-between"
    >
      <Heading>{heading}</Heading>
      <Divider size="small" />
      {counters.map(({ name, ...counterProps }, index) => (
        <Fragment key={buildKeyFromStrings(name, index)}>
          <FlexContainer flexDirection="row" justifyContent="space-between">
            <Paragraph>{name}</Paragraph>
            <Counter name={name} {...counterProps} />
          </FlexContainer>
          <Divider size="small" />
        </Fragment>
      ))}
    </FlexContainer>
  </Segment>
);

Component.displayName = 'CounterInputSegment';

Component.propTypes = {
  /**	An array of [`Counter`](#/Counter) props objects. */
  counters: PropTypes.arrayOf(PropTypes.object).isRequired,
  /** The text for the input segment's heading. */
  heading: PropTypes.string.isRequired,
};
