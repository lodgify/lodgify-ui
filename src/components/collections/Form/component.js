import React, { PureComponent, Children } from 'react';
import PropTypes from 'prop-types';
import { Card, Form, Message } from 'semantic-ui-react';
import { size, forEach } from 'lodash';

import { Button } from 'elements/Button';
import { Heading } from 'typography/Heading';
import { Link } from 'elements/Link';
import { SEND } from 'utils/default-strings';

import { getEmptyState } from './utils/getEmptyState';
import { getValidationWithDefaults } from './utils/getValidationWithDefaults';
import { getIsRequiredErrorThenSetState } from './utils/getIsRequiredErrorThenSetState';
import { getIsValidError } from './utils/getIsValidError';
import { getEmptyRequiredInputs } from './utils/getEmptyRequiredInputs';
import { getFormChild } from './utils/getFormChild';
import { getIsSubmitButtonDisabled } from './utils/getIsSubmitButtonDisabled';

/**
 * A form displays a set of related user input fields in a structured way.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export class Component extends PureComponent {
  state = {};

  componentDidUpdate = previousProps => {
    if (!previousProps.successMessage && !!this.props.successMessage) {
      const emptyState = getEmptyState(this.state);

      this.setState(emptyState);
    }
  };

  handleInputBlur = name =>
    getIsRequiredErrorThenSetState(
      this.setState.bind(this),
      this.props.validation[name],
      name,
      this.state[name]
    );

  handleInputChange = (name, value) => {
    const hasIsRequiredError = getIsRequiredErrorThenSetState(
      this.setState.bind(this),
      this.props.validation[name],
      name,
      { value }
    );

    if (hasIsRequiredError) {
      return;
    }

    const { invalidMessage, getIsValid } = getValidationWithDefaults(
      this.props.validation[name]
    );
    const isValid = getIsValid(value);
    const error = getIsValidError(value, isValid, invalidMessage);

    this.setState({ [name]: { error, isValid, value } });
  };

  handleSubmit = () => {
    const emptyRequiredInputs = getEmptyRequiredInputs(
      this.props.validation,
      this.state
    );

    size(emptyRequiredInputs)
      ? forEach(emptyRequiredInputs, (validation, name) => {
          const { isRequiredMessage } = getValidationWithDefaults(validation);

          this.setState({
            [name]: { error: isRequiredMessage },
          });
        })
      : this.props.onSubmit(this.state);
  };

  render = () => {
    const {
      actionLink,
      autoComplete,
      children,
      errorMessage,
      headingText,
      submitButtonText,
      successMessage,
    } = this.props;

    return (
      <Card className="has-form" fluid>
        {headingText && (
          <Card.Content>
            <Heading>{headingText}</Heading>
          </Card.Content>
        )}
        <Card.Content>
          <Form
            autoComplete={autoComplete}
            error={!!errorMessage}
            onSubmit={this.handleSubmit}
            success={!!successMessage}
          >
            {Children.map(children, child => getFormChild(child, this))}
            {!!successMessage && <Message content={successMessage} success />}
            {!!errorMessage && <Message content={errorMessage} error />}
            {actionLink && (
              <Link onClick={actionLink.onClick}>{actionLink.text}</Link>
            )}
            <Button
              isDisabled={getIsSubmitButtonDisabled(this.state)}
              isPositionedRight
            >
              {submitButtonText || SEND}
            </Button>
          </Form>
        </Card.Content>
      </Card>
    );
  };
}

Component.displayName = 'Form';

Component.defaultProps = {
  autoComplete: null,
  errorMessage: '',
  headingText: null,
  onSubmit: Function.prototype,
  actionLink: null,
  submitButtonText: SEND,
  successMessage: '',
  validation: {},
};

Component.propTypes = {
  /** An optional action link. */
  actionLink: PropTypes.shape({
    /** The function to call when the secondary call to action is clicked. */
    onClick: PropTypes.func,
    /** The visible text for the secondary call to action */
    text: PropTypes.node.isRequired,
  }),
  /** Can inputs be completed automatically by the browser. See [MDN docs for `<form />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form). */
  autoComplete: PropTypes.oneOf(['on', 'off']),
  /** The child components and elements. */
  children: PropTypes.node.isRequired,
  /** The message to display when the form has an error */
  errorMessage: PropTypes.string,
  /** The text to display as a heading at the top of the form. */
  headingText: PropTypes.string,
  /** The function to call when the form is submitted
   *  @param {Object} values - The values of the inputs in the form.
   */
  onSubmit: PropTypes.func,
  /** The text to display on the submit button. */
  submitButtonText: PropTypes.string,
  /** The message to display when the form is successful */
  successMessage: PropTypes.string,
  /** Settings for validating inputs. */
  validation: PropTypes.objectOf(
    PropTypes.shape({
      /** A function which checks whether the input is empty. Defaults to check for falsy value.
       *  @param  {any}     value
       *  @return {boolean}
       */
      getIsEmpty: PropTypes.func,
      /** A function which checks `value` and returns `true` if valid, `false` if invalid.
       *  @param  {any}     value
       *  @return {boolean}
       */
      getIsValid: PropTypes.func,
      /** An optional message to display if `getIsValid` returns `false`. */
      invalidMessage: PropTypes.string,
      /** Is the user required to enter a value into the input. */
      isRequired: PropTypes.bool,
      /** An optional message to display if `getIsValid` returns `false`. */
      isRequiredMessage: PropTypes.string,
    })
  ),
};
