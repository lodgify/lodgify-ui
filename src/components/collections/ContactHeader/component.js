import React from 'react';
import { Menu } from 'semantic-ui-react';
import {
  string,
  arrayOf,
  shape,
  oneOfType,
  oneOf,
  number,
  bool,
  func,
} from 'prop-types';

import { FlexContainer } from 'layout/FlexContainer';
import { Button } from 'elements/Button';
import { Submenu } from 'elements/Submenu';
import { size } from 'utils/size';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { ShowOn } from 'layout/ShowOn';
import { testidFactory } from 'utils/testid';
import { Logo } from 'elements/Logo';

import { HorizontalMenu } from '../HorizontalMenu';

import { MobileHeader } from './MobileHeader';
import { PHONE_PREFIX } from './constants';

const TEST_ID_PREFIX = 'contactHeader';

const testid = testidFactory(TEST_ID_PREFIX);

/**
 * A header displays a logo, grouped navigation items
 * contact information and an a primary call to action.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  primaryButtonText,
  navigationMenuItems,
  languageOptions,
  onChangeLanguage,
  currencyOptions,
  onChangeCurrency,
  currencyValue,
  currencyNoResultsText,
  languageValue,
  phoneNumber,
  logoHref,
  logoSubText,
  logoSize,
  logoSrc,
  logoSizes,
  logoText,
  logoSrcSet,
}) => (
  <header className="contact-header" data-testid={testid()}>
    <ShowOn computer tablet widescreen>
      <FlexContainer alignItems="center" justifyContent="space-between">
        <Menu>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoSubTextAlignment="center"
          logoText={logoText}
        />
        <Button>{primaryButtonText}</Button>
      </FlexContainer>
      <HorizontalMenu isHeader items={navigationMenuItems} />
    </ShowOn>
    <ShowOn mobile>
      <FlexContainer
        alignItems="center"
        className="hidden-contact-header"
        justifyContent="space-between"
      >
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
        />
        <MobileHeader
          currencyNoResultsText={currencyNoResultsText}
          currencyOptions={currencyOptions}
          currencyValue={currencyValue}
          languageOptions={languageOptions}
          languageValue={languageValue}
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
          navigationItems={navigationMenuItems}
          onChangeCurrency={onChangeCurrency}
          phoneNumber={phoneNumber}
        />
      </FlexContainer>
    </ShowOn>
  </header>
);

Component.displayName = 'ContactHeader';

Component.defaultProps = {
  currencyNoResultsText: undefined,
  currencyOptions: [],
  currencyValue: null,
  languageOptions: [],
  languageValue: null,
  logoHref: '',
  logoSize: 'medium',
  logoSizes: undefined,
  logoSrc: undefined,
  logoSrcSet: undefined,
  logoSubText: '',
  logoText: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
  phoneNumber: '',
  primaryButtonText: '',
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The href for the logo link. */
  logoHref: string,
  /** The maximum size of the headers logo. */
  logoSize: oneOf(['medium', 'large', 'huge']),
  /** A list of one or more strings separated by commas indicating a set of source sizes for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSizes: string,
  /** The src url for the logo. */
  logoSrc: string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSrcSet: string,
  /** The text that appears under the logo or logo text. */
  logoSubText: string,
  /** The text for the logo. */
  logoText: string,
  /** The items for a user to navigate the site. */
  navigationMenuItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
  /** The text to display for the primary button. */
  primaryButtonText: string,
};
