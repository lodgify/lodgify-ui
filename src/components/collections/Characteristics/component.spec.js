import React from 'react';
import { shallow } from 'enzyme';
import { expectComponentToHaveDisplayName } from '@lodgify/enzyme-jest-expect-helpers';

import { propertyCharacteristics } from './mock-data/props';
import { Component as Characteristics } from './component';

const props = {
  spaceCharacteristics: propertyCharacteristics,
};

const getCharacteristics = props => shallow(<Characteristics {...props} />);

describe('<Characteristics />', () => {
  it('should have the correct structure', () => {
    const wrapper = getCharacteristics(props);

    expect(wrapper).toMatchSnapshot();
  });
  it('should have the correct structure', () => {
    const wrapper = getCharacteristics({ propertyCharacteristics: [] });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have `displayName` `Characteristics`', () => {
    expectComponentToHaveDisplayName(Characteristics, 'Characteristics');
  });
});
