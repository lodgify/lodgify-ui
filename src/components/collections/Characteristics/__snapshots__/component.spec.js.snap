// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Characteristics /> should have the correct structure 1`] = `
<Characteristics
  propertyCharacteristics={
    Array [
      Object {
        "iconName": "users",
        "text": "2 guests",
      },
      Object {
        "iconName": "home",
        "text": "1 bedroom",
      },
      Object {
        "iconName": "single bed",
        "text": "2 beds",
      },
      Object {
        "iconName": "bathroom",
        "text": "1 bathroom",
      },
    ]
  }
  spaceCharacteristics={Array []}
>
  <Grid
    areColumnsCentered={false}
    columns={1}
    isStackable={false}
  >
    <Grid
      centered={false}
      columns={1}
      stackable={false}
    >
      <div
        className="ui one column grid"
      >
        <GridColumn
          verticalAlignContent="top"
          width={12}
        >
          <GridColumn
            verticalAlign="top"
            width={12}
          >
            <div
              className="top aligned twelve wide column"
            >
              <ShowOn
                computer={true}
                mobile={false}
                parent={[Function]}
                parentProps={
                  Object {
                    "horizontal": true,
                  }
                }
                tablet={true}
                widescreen={true}
              >
                <List
                  className="show-on-tablet show-on-computer show-on-widescreen"
                  horizontal={true}
                >
                  <div
                    className="ui horizontal list show-on-tablet show-on-computer show-on-widescreen"
                    role="list"
                  />
                </List>
              </ShowOn>
              <ShowOn
                computer={false}
                mobile={true}
                parent={[Function]}
                parentProps={
                  Object {
                    "columns": 1,
                  }
                }
                tablet={false}
                widescreen={false}
              >
                <Grid
                  areColumnsCentered={false}
                  className="show-on-mobile"
                  columns={1}
                  isStackable={false}
                >
                  <Grid
                    centered={false}
                    className="show-on-mobile"
                    columns={1}
                    stackable={false}
                  >
                    <div
                      className="ui one column grid show-on-mobile"
                    />
                  </Grid>
                </Grid>
              </ShowOn>
            </div>
          </GridColumn>
        </GridColumn>
      </div>
    </Grid>
  </Grid>
</Characteristics>
`;
